### ISSUE TITLE
"Enable Hotkey" doesn't work across multiple devices

### ISSUE THREAD

**Issue by BenMcLean** [libretro/RetroArch#15492]
## Description

I'm trying to set up custom controls for the non-Steam RetroArch on the Steam Deck and have run into this RetroArch bug that's giving me a real hard time.

### Expected behavior

If I set "Enable Hotkeys" to a keyboard key, then none of the hotkeys should activate when I am not holding that key down, (they should only activate when I am holding that key down) regardless of what control they are mapped to, no matter which control device.

### Actual behavior

If I set "Enable Hotkeys" to a keyboard key, but some of the hotkey functions are mapped to the game controller, the ones mapped to the game controller don't require the keyboard key to be pressed. They activate regardless whether the keyboard key is pressed or not.

### Steps to reproduce the bug

1. Set "Enable Hotkeys" to a keyboard key, such as Q
2. Set "Save State" to a controller button, such as right shoulder.
3. Press right shoulder in game without pressing Q and you'll see the save state function activated, even though it shouldn't be activated without pressing Q.

### Bisect Results

The concept of a separate keyboard and joypad are sort of made obsolete by the Steam Deck and other handhelds that have more buttons than an XBox controller, so naturally one of those "extra" buttons that I'm trying to use for "Enable Hotkeys" would have to get mapped to the keyboard. Except I can't do that because of this bug.

### Version/Commit
6616b80 built on May 16 2023, flatpak downloaded through Discover Software Center on the Steam Deck.

### Environment information

- OS: Steam OS on the Steam Deck
- Compiler: GCC (12.2.0 64-bit)

This issue is similar to #14048 because changing that would also solve this I think.
I also posted about this on the forums: https://forums.libretro.com/t/enable-hotkeys-bug-or-just-unexpected-behavior/42140

**Comment by hizzlekizzle** [comment]
This is actually intended behavior. The hotkey enabler will only affect the "type" of device that it was mapped onto. Now, you can still map multiple types of devices (e.g., a key on your keyboard and a button on your gamepad simultaneously) and they will each affect their device type's hotkeys invocations.

**Comment by BenMcLean** [comment]
> This is actually intended behavior.

This is awful behavior since the Settings -> Input -> Hotkeys menu will let you map anything and won't tell you that the "Enable Hotkeys" key only works on the one device you can set it to, so the result is that it appears that some mappings just randomly ignore the "Enable Hotkeys" key and some don't

> Now, you can still map multiple types of devices (e.g., a key on your keyboard and a button on your gamepad simultaneously) and they will each affect their device type's hotkeys invocations.

But then you have to sacrifice one of your precious XBox 360 controller buttons that games actually use to this crap when you shouldn't ever have to do that.

**Comment by hizzlekizzle** [comment]
Yeah, it's a fairly recent change that was implemented because if someone mapped a hotkey to, say, a keyboard to streamline using a classic computer core, it prevented them from using any hotkeys on their gamepad unless they walked across the room or whatever and held down the key on their keyboard.

I'll see about adding some sublabel text that explains the per-device-type nature.

**Comment by BenMcLean** [comment]
> if someone mapped a hotkey to, say, a keyboard to streamline using a classic computer core, it prevented them from using any hotkeys on their gamepad unless they walked across the room or whatever and held down the key on their keyboard.

I don't understand how doing it this way makes that particular situation better. It forces everyone, on every type of gamepad, to sacrifice one of their buttons to "Enable Hotkeys" if they want any of their buttons to be involved in a hotkey combination.

> you can still map multiple types of devices (e.g., a key on your keyboard and a button on your gamepad simultaneously) and they will each affect their device type's hotkeys invocations.

I did work out a sort of compromise for this. I mapped L3 on the Steam Deck to BOTH the `[` keyboard key AND the right joystick click. Then I set the "Enable Hotkeys" in RetroArch to map to BOTH of those controls. That way, one button enables hotkeys for both devices.
However, this means I have sacrificed right stick click. Games can't use it now. I shouldn't have to do that.

**Comment by sonninnos** [comment]
The previous behavior was way more awful, since it meant having to use both controller and keyboard if either has the hotkey set. Keyboard hotkeys would be completely unusable without holding a controller while using the keyboard..

**Comment by BenMcLean** [comment]
> The previous behavior was way more awful, since it meant having to use both controller and keyboard if either has the hotkey set. Keyboard hotkeys would be completely unusable without holding a controller while using the keyboard..

Two things to say about this:
1. I don't want a delay because I want to be able to get to the hotkey functions quick, plus I don't want the game to get that button press before the delay
2. Couldn't we get a toggle for the old behavior vs the new behavior? The old behavior would be fine for the Steam Deck.

**Comment by bradrhodes** [comment]
I'm really hoping that the devs can take another look at their approach for fixing this problem. This has made making hotkeys on devices like steam deck unnecessarily convoluted and completely unintuitive. I've spent probably 10 hours trying to come up with a hotkey solution that works for me now. It used to work flawlessly, then I updated retroarch. I spent hours googling to try to figure out why my hotkeys no longer worked (or rather it was unloading because it didn't require holding the enable button down) just to find this thread to explain that the devs made a breaking change. Poor breaking change management first of all. Second of all, could we just have a setting for classic behavior? And that should be enabled by default in order to manage a breaking change properly.
